### PR TITLE
Allow use of Redis for all caches

### LINKED ISSUE #4341: Async cache writes use request scoped context which gets cancelled
#### Describe the bug

Stores to caches use the [request context](https://github.com/grafana/mimir/blob/4101955b5dc8e19210f023ae92bdf671a46ae40a/pkg/storegateway/bucket_index_reader.go#L466). However, the underlying cache wrappers in dskit `cache.Cache` and `cache.RemoteCacheClient` store things asynchronously via a job queue. This means that stores to the cache might still be running when the request ends. Because of this, cache clients that respect contexts (e.g. the Redis client that will be introduced in #4236) will fail to store things.

#### To Reproduce

N/A

#### Expected behavior

Stores should use a background context so that they aren't aborted when a request ends.

#### Environment

- Infrastructure: [e.g., Kubernetes, bare-metal, laptop]
- Deployment tool: [e.g., helm, jsonnet]

### PR DESCRIPTION
**Pull request by 56quarters**
#### What this PR does

This change pulls in a version of dskit that supports using Redis implementations for the `cache.Cache` and `cache.RemoteCacheClient` interfaces. The query-frontend caching integration test now runs with both Memcached and Redis caches. Redis can also be enabled for the microservices docker compose environment.

#### Which issue(s) this PR fixes or relates to

See #4236
Fixes #4341

#### Checklist

- [X] Tests updated
- [X] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by 56quarters** [comment]
Note that I am purposefully not including updates to the Jsonnet or Helm chart as part of these changes (there's enough to review as-is).

File: cmd/mimir/help.txt.tmpl [hunk 1]
```diff
@@ -432,13 +456,21 @@ Usage of ./cmd/mimir/mimir:
   -query-frontend.query-sharding-total-shards int
     	The amount of shards to use when doing parallelisation via query sharding by tenant. 0 to disable query sharding for tenant. Query sharding implementation will adjust the number of query shards based on compactor shards. This allows querier to not search the blocks which cannot possibly have the series for given query shard. (default 16)
   -query-frontend.results-cache.backend string
-    	Backend for query-frontend results cache, if not empty. Supported values: [memcached].
+    	Backend for query-frontend results cache, if not empty. Supported values: [memcached redis].
```
Comment (charleskorn) [inline cmd/mimir/help.txt.tmpl:459]: [nit] The other descriptions have it formatted like this:
```suggestion
    	Backend for query-frontend results cache, if not empty. Supported values: memcached, redis.
```

File: pkg/storage/tsdb/index_cache.go [hunk 2]
```diff
@@ -119,3 +125,17 @@ func newMemcachedIndexCache(cfg cache.MemcachedConfig, logger log.Logger, regist
 
 	return indexcache.NewTracingIndexCache(c, logger), nil
 }
+
+func newRedisIndexCache(cfg cache.RedisClientConfig, logger log.Logger, registerer prometheus.Registerer) (indexcache.IndexCache, error) {
+	client, err := cache.NewRedisClient(logger, "index-cache", cfg, prometheus.WrapRegistererWithPrefix("thanos_", registerer))
```
Comment (lamida) [inline pkg/storage/tsdb/index_cache.go:130]: We had a discussion [to deprecate `thanos_` metrics prefix](https://github.com/grafana/mimir/issues/4241). Should we just use `cortex_` prefix for the new change? Applicable also for memcached client metrics prefix few lines above.

Comment (56quarters) [reply pkg/storage/tsdb/index_cache.go:130]: I'd rather make the change to all cache metrics at the same time. It's the same amount of work if we support `cortex_` for Redis metrics and `thanos_` for Memcached vs using `thanos_` for both now and `cortex_` for both later.

File: cmd/mimir/help.txt.tmpl (hunk 1 above)
Comment (56quarters) [reply cmd/mimir/help.txt.tmpl]: Fixed!

File: pkg/storage/tsdb/index_cache.go (hunk 2 above)
Comment (56quarters) [reply pkg/storage/tsdb/index_cache.go:130]: To be more specific, using `thanos_` here now means Redis will work with our dashboards as currently written with no extra effort. Then, they can be migrated to use the `cortex_` prefix at the same time as the Memcached backend also with no extra effort.

**Review by pracucci — APPROVED** [review]
Very good job, LGTM! I just left a couple of comments I would be glad if you could look at before merging. I would love to merge this PR asap, cause I want to add metadata cache support to compactor and I would love to do it on top of this. Thanks!

File: pkg/storage/tsdb/index_cache.go [hunk 3]
```diff
@@ -52,8 +55,9 @@ func (cfg *IndexCacheConfig) RegisterFlags(f *flag.FlagSet) {
...
 	f.StringVar(&cfg.Backend, prefix+"backend", IndexCacheBackendDefault, fmt.Sprintf("The index cache backend type. Supported values: %s.", strings.Join(supportedIndexCacheBackends, ", ")))
 
-	cfg.InMemory.RegisterFlagsWithPrefix(f, prefix+"inmemory.")
-	cfg.Memcached.RegisterFlagsWithPrefix(f, prefix+"memcached.")
+	cfg.InMemory.RegisterFlagsWithPrefix(prefix+"inmemory.", f)
+	cfg.Memcached.RegisterFlagsWithPrefix(prefix+"memcached.", f)
+	cfg.Redis.RegisterFlagsWithPrefix(prefix+"redis.", f)
 }
 
 // Validate the config.
```
[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storage/tsdb/index_cache.go:63]: D'oh, will do.

[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storage/tsdb/index_cache.go:63]: Exposing the errors here so I can write a test against the validation method: https://github.com/grafana/dskit/pull/276

**Pull request merged**